Mathlib/Combinatorics/SimpleGraph/Tutte.lean

```diff
@@ -0,0 +1,125 @@
+    rw [hM1.2.support_eq_verts, hM2.support_eq_verts, Subgraph.verts_iSup,
+      Set.disjoint_iUnion_right]
+    exact fun K ↦ (compMatching K).choose_spec.1.symm ▸ Set.disjoint_sdiff_right
+  -- The only vertices left are indeed contained in universalVerts
+  have : (M1.verts ∪ M2.verts)ᶜ ⊆ G.universalVerts := by
+    rw [Set.compl_subset_comm, Set.compl_eq_univ_diff]
+    intro v hv
+    by_cases h : v ∈ M1.verts
+    · exact M1.verts.mem_union_left _ h
+    right
+    simp only [deleteUniversalVerts_verts, Subgraph.verts_iSup, Set.mem_iUnion, M2,
+      (compMatching _).choose_spec.1]
+    use (G.deleteUniversalVerts.coe.connectedComponentMk ⟨v, hv⟩)
+    aesop
+  exact ⟨M1 ⊔ M2, hM1.2.sup hM2 disjointM12, this⟩
+
+/-- A graph in which the universal vertices do not violate the
+Tutte-condition, if the graph decomposes into cliques, it has a perfect matching. -/
+theorem Subgraph.IsPerfectMatching.exists_of_isClique_supp
+    (hveven : Even (Fintype.card V)) (h : ¬G.IsTutteViolator G.universalVerts)
+    (h' : ∀ (K : G.deleteUniversalVerts.coe.ConnectedComponent),
+    G.deleteUniversalVerts.coe.IsClique K.supp) : ∃ (M : Subgraph G), M.IsPerfectMatching := by
+  classical
+  obtain ⟨M, ⟨hM, hsub⟩⟩ := IsMatching.exists_verts_compl_subset_universalVerts h h'
+  obtain ⟨M', hM'⟩ := ((G.isClique_universalVerts.subset hsub).even_iff_exists_isMatching
+    (Set.toFinite _)).mp (by simpa [Set.even_ncard_compl_iff hveven, -Set.toFinset_card,
+      ← Set.ncard_eq_toFinset_card'] using hM.even_card)
+  use M ⊔ M'
+  have hspan : (M ⊔ M').IsSpanning := by
+    rw [Subgraph.isSpanning_iff, Subgraph.verts_sup, hM'.1]
+    exact M.verts.union_compl_self
+  exact ⟨hM.sup hM'.2 (by
+    simp only [hM.support_eq_verts, hM'.2.support_eq_verts, hM'.1, Subgraph.verts_sup]
+    exact (Set.disjoint_compl_left_iff_subset.mpr fun ⦃a⦄ a ↦ a).symm), hspan⟩
+
+theorem IsTutteViolator.empty (hodd : Odd (Fintype.card V)) : G.IsTutteViolator ∅ := by
+  classical
+  have ⟨c, hc⟩ := Classical.inhabited_of_nonempty
+    (Finite.card_pos_iff.mp <| Odd.pos <|
+    (odd_ncard_oddComponents ((⊤ : Subgraph G).deleteVerts ∅).coe).mpr (by
+    simpa [Fintype.card_congr (Equiv.Set.univ V)] using hodd))
+  rw [IsTutteViolator, Set.ncard_empty, Set.ncard_pos]
+  use c
+
+/-- Proves the necessity part of Tutte's theorem -/
+lemma not_isTutteViolator_of_isPerfectMatching {M : Subgraph G} (hM : M.IsPerfectMatching)
+    (u : Set V) :
+    ¬G.IsTutteViolator u := by
+  simpa [IsTutteViolator, Set.Nat.card_coe_set_eq] using Finite.card_le_of_injective
+      (fun c => ⟨(ConnectedComponent.odd_matches_node_outside hM c).choose,
+        (ConnectedComponent.odd_matches_node_outside hM c).choose_spec.1⟩) (by
+    intro x y hxy
+    obtain ⟨v, hv⟩ := (ConnectedComponent.odd_matches_node_outside hM x).choose_spec.2
+    obtain ⟨w, hw⟩ := (ConnectedComponent.odd_matches_node_outside hM y).choose_spec.2
+    obtain ⟨v', hv'⟩ := (M.isPerfectMatching_iff).mp hM _
```

Comment:
```suggestion
    obtain ⟨v', hv'⟩ := M.isPerfectMatching_iff.mp hM _
```

Reply:
I replaced this with Iff.mp, the .mp notation gives an error: 
```
function expected at
  Subgraph.isPerfectMatching_iff
term has type
  Subgraph.IsPerfectMatching ?m.55222 ↔ ∀ (v : ?m.55220), ∃! w, Subgraph.Adj ?m.55222 v w
```

I'm not sure what's going on here, do you think it's worth minimizing this?

The inferred term is `(ConnectedComponent.odd_matches_node_outside hM y).choose`